synapse/handlers/space_summary.py

```diff
@@ -181,34 +190,42 @@ async def get_space_summary(
 
                     # All rooms returned don't need visiting again (even if the user
                     # didn't have access to them).
-                    processed_rooms.add(fed_room_id)
-
-                for event in fed_events:
-                    if event.get("room_id") in room_ids:
-                        events.append(event)
+                    processed_rooms[fed_room_id] = include_room
 
                 logger.debug(
                     "Query of %s returned rooms %s, events %s",
                     room_id,
                     [room.get("room_id") for room in fed_rooms],
-                    ["%s->%s" % (ev["room_id"], ev["state_key"]) for ev in fed_events],
+                    ["%s->%s" % (ev["room_id"], ev["state_key"]) for ev in events],
                 )
 
-            # the room we queried may or may not have been returned, but don't process
-            # it again, anyway.
-            processed_rooms.add(room_id)
+            # There might be events which point to this room which are waiting
+            # to see if it is accessible.
+            room_pending_events = pending_events.pop(room_id, ())
+            if not processed_rooms[room_id]:
+                room_pending_events = ()
 
-            # XXX: is it ok that we blindly iterate through any events returned by
-            #   a remote server, whether or not they actually link to any rooms in our
-            #   tree?
-            for ev in events:
-                events_result.append(ev)
+            # Return the events which reference rooms that were found to be
+            # accessible. Otherwise, queue them until we process those rooms.
+            for ev in itertools.chain(events, room_pending_events):
+                parent_room_id = ev["room_id"]
+                child_room_id = ev["state_key"]
+
+                try:
+                    if (
+                        processed_rooms[parent_room_id]
+                        and processed_rooms[child_room_id]
+                    ):
+                        events_result.append(ev)
+                except KeyError:
+                    # Note that events are pending of the child event since if
+                    # the parent event was not accessible it shouldn't be included
+                    # at all.
```

Comment:
I don't really follow this logic. Why do we only add the event to the pending list for the child room?

Reply:
The thought process was that we only want to send the event in the case of both the parent and child room being accessible to the requester.

I think the assumption here is that if the parent room isn't in `processed_rooms` something terrible has happened? There are two sources of data we're iterating over:

* `events` - `m.space.child` events from the current room.
	* For local rooms, the parent will be the current room, the child will be a room that may or may not have been processed.
	* For remote rooms, this includes the above, as well as additional events that might be children of children, etc. (Or even events that aren't part of the graph.)
* `room_pending_events` - `m.space.child` events which were previously found to be pointing to the current room, i.e. the child room is the current room and the parent room is a previously processed room.

I'm now having trouble convincing myself that parents must be processed before children and I think that assumption might be bogus actually.